File: src/connectors/emailrep.js
import { BaseConnector } from './base.js';
import https from 'https';

export class EmailRepConnector extends BaseConnector {
    constructor(config, state, telemetry) {
        super(config, state, telemetry);
        this.name = 'EmailRep';
    }

    async run(investigationId, phaseId, inputType, inputValue) {
        const runId = this.state.recordConnectorRun(investigationId, phaseId, this.name, inputType, inputValue);
        this.telemetry.connectorStart(investigationId, this.name, phaseId, { type: inputType, value: inputValue });

        try {
            const entities = [];
            const data = JSON.parse(await this._fetch(`https://emailrep.io/${encodeURIComponent(inputValue)}`));

            if (data.email) {
                entities.push({
                    type: 'email',
                    data: {
                        address: data.email,
                        reputation: data.reputation,
                        suspicious: data.suspicious,
                        references: data.references,
                        details: {
                            blacklisted: data.details?.blacklisted,
                            malicious_activity: data.details?.malicious_activity,
                            credential_leaked: data.details?.credentials_leaked,
                            data_breach: data.details?.data_breach,
                            first_seen: data.details?.first_seen,
                            last_seen: data.details?.last_seen,
                            domain_exists: data.details?.domain_exists,
                            domain_reputation: data.details?.domain_reputation,
                            new_domain: data.details?.new_domain,
                            days_since_domain_creation: data.details?.days_since_domain_creation,
                            spam: data.details?.spam,
                            free_provider: data.details?.free_provider,
                            deliverable: data.details?.deliverable,
                            accepted_all: data.details?.accepted_all,
                            valid_mx: data.details?.valid_mx,
                            spoofable: data.details?.spoofable,
                            spf_strict: data.details?.spf_strict,
                            dmarc_enforced: data.details?.dmarc_enforced
                        },
                        profiles: data.details?.profiles || [],
                        source: 'emailrep'
                    },
                    confidence: 0.85
                });

                if (data.details?.profiles?.length) {
                    for (const profile of data.details.profiles) {
                        entities.push({
                            type: 'account',
                            data: {
                                platform: profile,
                                username: null,
                                email_registered: true,
                                source: 'emailrep'
                            },
                            confidence: 0.7
                        });
                    }
                }

                if (data.details?.credentials_leaked || data.details?.data_breach) {
                    entities.push({
                        type: 'breach',
                        data: {
                            name: 'emailrep_detected',
                            email: inputValue,
                            credential_leaked: data.details?.credentials_leaked,
                            data_breach: data.details?.data_breach,
                            source: 'emailrep'
                        },
                        confidence: 0.7
                    });
                }
            }

            let newCount = 0;
            for (const entity of entities) {
                const added = this.state.addEntity(investigationId, entity.type, entity.data, entity.confidence || 0.5);
                if (added.new) newCount++;
                this.telemetry.entityDiscovered(investigationId, entity.type, added.new, this.name);
            }

            this.state.completeConnectorRun(runId, 'completed', newCount, null, null, 0);
            this.telemetry.connectorEnd(investigationId, this.name, phaseId, {
                status: 'completed', entitiesFound: newCount, input: { type: inputType, value: inputValue }
            });
            return { status: 'completed', entities, newCount };
        } catch (e) {
            this.state.completeConnectorRun(runId, 'failed', 0, null, e.message, 1);
            this.telemetry.connectorEnd(investigationId, this.name, phaseId, {
                status: 'failed', entitiesFound: 0, input: { type: inputType, value: inputValue }
            });
            return { status: 'failed', error: e.message };
        }
    }

    _fetch(url) {
        return new Promise((resolve, reject) => {
            const opts = { headers: { 'User-Agent': 'cyclops-osint', 'Accept': 'application/json' }, timeout: 15000 };
            const key = process.env.EMAILREP_API_KEY;
            if (key) opts.headers['Key'] = key;
            https.get(url, opts, res => {
                let data = '';
                res.on('data', chunk => { data += chunk; });
                res.on('end', () => resolve(data));
            }).on('error', reject);
        });
    }

    async healthCheck() { return { ok: true, version: 'native' }; }
}

File: cyclops/connectors/emailrep.py
import os
import aiohttp
from cyclops.connectors.base import BaseConnector


class EmailRepConnector(BaseConnector):
    def __init__(self, config, state, telemetry):
        super().__init__(config, state, telemetry)
        self.name = "EmailRep"
        self.api_key = os.environ.get("EMAILREP_API_KEY")

    async def run(self, investigation_id, phase_id, input_type, input_value):
        run_id = self.state.record_connector_run(investigation_id, phase_id, self.name, input_type, input_value)
        self.telemetry.connector_start(investigation_id, self.name, phase_id, {"type": input_type, "value": input_value})
        try:
            headers = {"User-Agent": "CYCLOPS-OSINT", "Accept": "application/json"}
            if self.api_key:
                headers["Key"] = self.api_key
            entities = []
            async with aiohttp.ClientSession(headers=headers) as session:
                async with session.get(f"https://emailrep.io/{input_value}", timeout=aiohttp.ClientTimeout(total=10)) as resp:
                    if resp.status == 404:
                        pass
                    elif resp.status >= 400:
                        raise Exception(f"EmailRep HTTP {resp.status}")
                    else:
                        data = await resp.json()
                        entities.append({"type": "email", "data": {
                            "address": input_value, "reputation": data.get("reputation"), "suspicious": data.get("suspicious"),
                            "references": data.get("references"), "details": data.get("details"), "source": "emailrep",
                        }, "confidence": 0.8})
                        profiles = data.get("details", {}).get("profiles", [])
                        for p in profiles:
                            entities.append({"type": "account", "data": {"platform": p, "address": input_value, "source": "emailrep"}, "confidence": 0.7})
            new_count = 0
            for entity in entities:
                added = self.state.add_entity(investigation_id, entity["type"], entity["data"], entity.get("confidence", 0.5))
                if added["new"]:
                    new_count += 1
                self.telemetry.entity_discovered(investigation_id, entity["type"], added["new"], self.name)
            self.state.complete_connector_run(run_id, "completed", new_count)
            self.telemetry.connector_end(investigation_id, self.name, phase_id, {"status": "completed", "entitiesFound": new_count, "input": {"type": input_type, "value": input_value}})
            return {"status": "completed", "entities": entities, "newCount": new_count}
        except Exception as e:
            self.state.complete_connector_run(run_id, "failed", 0, error=str(e))
            self.telemetry.connector_end(investigation_id, self.name, phase_id, {"status": "failed", "input": {"type": input_type, "value": input_value}})
            return {"status": "failed", "error": str(e)}

    async def health_check(self):
        return {"ok": True, "hasApiKey": bool(self.api_key)}
